src/core/global_include.h

```diff
@@ -63,6 +67,32 @@ class FakeTime : public Time {
     void add_overhead();
 };
 
+class FCUTime {
+public:
+    FCUTime();
+    virtual ~FCUTime();
+
+    dl_fcu_time_t now();
+
+    template<typename T> void shift_fcu_time_by(T offset)
+    {
+        std::lock_guard<std::mutex> lock(_fcu_system_time_offset_mutex);
+        _fcu_time_offset = std::chrono::duration_cast<std::chrono::nanoseconds>(offset);
+    };
+
+    template<typename T> inline dl_fcu_time_t time_in(T local_system_time_point)
+    {
+        std::lock_guard<std::mutex> lock(_fcu_system_time_offset_mutex);
+        return dl_fcu_time_t(local_system_time_point + std::chrono::duration_cast<T>(_fcu_time_offset));
+    };
+
+private:
+    mutable std::mutex _fcu_system_time_offset_mutex{};
+    std::chrono::nanoseconds _fcu_time_offset{};
+
+    virtual dl_stime_t system_time();
+};
```

Comment:
Does this all need to be globally available? I guess so for the templates?

Reply:
Refactored